### PR TITLE
Json format macro type fixes

### PR DESCRIPTION
**Pull request by theo-lanman**
Here are two fixes to the play json "format" macro.

#### 0e43701: Re-check recursive-type filter after applying bound type
Recursive types (e.g., `case class Foo(a: Int, b: Option[Foo])` are implemented by replacing references to the outer type with a placeholder. Generic types have an interesting property (at macro-time) that references to the type parameter are not yet bound to the parameter—the type references aren't equal. There is one function that walks the type tree and both A) replaces references to the type parameter (by name) with the value of the type parameter (https://github.com/semanticmachines/playframework/commit/0e43701b97c7e671c456fac8b46b426b4a6d5cbb#diff-affc82d0f265c1cd61d0add6d17c48c1R296), and B) replaces instances of the type in its own tree with the placeholder reference. This doesn't happen in the correct order, which causes generic, recursive types to fail. This commit fixes this somewhat heavy-handedly.

#### f8d7e51: Scala does not eta-expand overloaded generics; do it "by hand"
Scala won't eta-expand generic methods:
```scala
def m[T](x: T): T = x
def m[T](x: T, y: T): T = y
val f = m[Int] _
// error: _ must follow method; cannot follow (x: Int, y: Int)Int <and> (x: Int)Int
```

(Note: adding the appropriate type ascription works in the REPL, but it doesn't work in a macro.)

Thankfully we're in macroland, where there's enough information to spell it out. This commit fixes this very heavy-handedly: this requires the same type-binding logic as mentioned in the previous fix, which has just been copy-pasted rather than refactored.

### CODE REVIEW

File: framework/src/play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala [hunk 1]
```diff
@@ -247,7 +247,37 @@ object JsMacroImpl {
               q"(..$vals) => $companionObject.apply(..${applyParams.init}, ${applyParams.last}: _*)"
             } else if (tpeArgs.isEmpty) {
               q"$companionObject.apply _"
-            } else q"$companionObject.apply[..$tpeArgs] _"
+            } else {
+              // TODO: refactor (see: boundTypes, refactor, normalized in ImplicitResolver
+              val boundTypes: Map[String, Type] = app.typeParams.map(_.asType.fullName).zip(tpeArgs).toMap
+              @annotation.tailrec
+              def refactor(in: List[Type], base: TypeSymbol, out: List[Type], tail: List[(List[Type], TypeSymbol, List[Type])]): Type = in match {
```
Comment (dlwh) [inline framework/src/play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala]: i'm having a lot of trouble following this method. can you add a note explaining it?

Comment (theo-lanman) [inline framework/src/play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala]: Added a comment; let me know if that's sufficient or if you want me to add more.

Comment (dlwh) [inline framework/src/play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala]: oh sorry didn't realize you copied it from elsewhere. i'm pleased.

**Comment by dlwh** [comment]
lgtm

**Pull request merged**